src/python/zquantum/core/circuit/conversions/pyquil_conversions.py

```diff
@@ -203,3 +221,51 @@ def convert_circuit_to_pyquil(
         program += convert_to_pyquil(gate, program)
 
     return program
+
+
+@singledispatch
+def convert_from_pyquil(obj: Union[pyquil.Program, pyquil.quil.Gate]):
+    pass
+
+
+@convert_from_pyquil.register
+def convert_gate_from_pyquil(gate: pyquil.quil.Gate) -> Gate:
+    number_of_control_modifiers = gate.modifiers.count("CONTROLLED")
+
+    all_qubits = pyquil_qubits_to_numbers(gate.qubits)
+    control_qubits = all_qubits[:number_of_control_modifiers]
+    original_qubits = all_qubits[number_of_control_modifiers:]
+
+    try:
+        gate_cls = PYQUIL_NAME_TO_ORQUESTRA_CLS[gate.name]
+        if (
+            gate_cls in SINGLE_QUBIT_NONPARAMETRIC_GATES
+            or gate_cls in TWO_QUBIT_CONTROLLED_GATES
+        ):
+            result = gate_cls(*original_qubits)
+        elif gate_cls in ROTATION_GATES or gate_cls == CPHASE:
+            result = gate_cls(
+                *original_qubits,
+                translate_expression(
+                    expression_from_pyquil(gate.params[0]), SYMPY_DIALECT
+                ),
+            )
+        elif gate_cls == SWAP:
+            result = gate_cls(original_qubits)
+        else:
+            raise RuntimeError(
+                f"Error converting gate {gate}. If you see this message, "
+                "please file a bugreport."
+            )
+
+        for qubit in reversed(control_qubits):
+            result = ControlledGate(result, qubit)
+
+        if gate.modifiers.count("DAGGER") % 2 == 1:
```

Comment:
This seems odd. Does `pyquil` support doing multiple `DAGGER` modifiers as opposed to just a toggle?

Reply:
Unfortunately yes, PyQuil allows applying multiple dagger modifiers. I suppose that optimizing them out is left to the Quil compiler later on so they didn't bother to implement it in PyQuil itself. To see that this is the case, consider the following code, which prints `["DAGGER", "DAGGER", "DAGGER"]`
```python
from pyquil.gates import X
gate = X(2).dagger().dagger().dagger()
print(gate.modifiers)
```

Reply:
I added a comment in d82c2a1 that describes why we test for number of DAGGER modifiers.